gui-agent/txrx-vchan.c

```diff
@@ -84,47 +82,37 @@ int read_data(libvchan_t *vchan, char *buf, int size)
     return size;
 }
 
-static int wait_for_vchan_or_argfd_once(libvchan_t *vchan, int nfd, int *fd, fd_set * retset)
+static int wait_for_vchan_or_argfd_once(libvchan_t *vchan, struct pollfd *const fds, size_t const nfds)
 {
-    fd_set rfds;
-    int vfd, max = 0, ret, i;
-    vfd = libvchan_fd_for_select(vchan);
-    FD_ZERO(&rfds);
-    for (i = 0; i < nfd; i++) {
-        int cfd = fd[i];
-        FD_SET(cfd, &rfds);
-        if (cfd > max)
-            max = cfd;
-    }
-    FD_SET(vfd, &rfds);
-    if (vfd > max)
-        max = vfd;
-    max++;
-    ret = select(max, &rfds, NULL, NULL, NULL);
-    if (ret < 0 && errno == EINTR)
-        return 0;
+    int ret;
+    write_data(vchan, NULL, 0);    // trigger write of queued data, if any present
+    ret = poll(fds, nfds, 1000);
     if (ret < 0) {
-        perror("select");
+        if (errno == EINTR)
+            return -1;
+        perror("poll");
         exit(1);
     }
     if (!libvchan_is_open(vchan)) {
         fprintf(stderr, "libvchan_is_eof\n");
+        libvchan_close(vchan);
```

Comment:
You introduce double-free here.
Please don't combine subtle behavior changes with refactors (select->poll in this case).